torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -751,24 +758,47 @@ void Reducer::initialize_buckets(
 // (see Note:  "Gradient Layout Contract" in initialize_buckets).
 void Reducer::initialize_bucketviews(
     Reducer::BucketReplica& replica,
-    at::Tensor& contents) {
+    at::Tensor& contents,
+    bool copy_to_bucket_view) {
   for (size_t i = 0; i < replica.variables.size(); i++) {
-    const auto& v = replica.variables[i];
+    auto& v = replica.variables[i];
     const auto offset = replica.offsets[i];
     const auto length = replica.lengths[i];
+    at::Tensor bucket_view;
     if (v.is_non_overlapping_and_dense()) {
       // If the param's memory is dense, match its layout, anticipating
       // the autograd engine (AccumulateGrad) will also create gradients
       // matching its layout.
-      replica.bucket_views.push_back(
-          contents.as_strided(v.sizes(), v.strides(), offset));
+      bucket_view = contents.as_strided(v.sizes(), v.strides(), offset);
     } else {
       // Fall back to a C-style contiguous view, again anticipating
       // AccumulateGrad will do the same when stashing grads for non-dense
       // params.
-      replica.bucket_views.push_back(
-          contents.narrow(0, offset, length).view(v.sizes()));
+      bucket_view = contents.narrow(0, offset, length).view(v.sizes());
     }
+    replica.bucket_views.push_back(bucket_view);
+    // If grad has already been defined/calculated in previous iterations,
+    // and the grad does not point to bucket_view, that means bucket_view
+    // is newly allocated, if this function is called inside initialize_bucket,
+    // old grad needs to be copied into new bucket_view for initialization
+    // and let grad point to bucket_view;
+    // otherwise this function is called for communication hook, bucket_view
+    // has the updated results in new tensor, just let grad point to
+    // bucket_view. If grad is not defined, do not let it point to bucket_view.
+    // E.g., for global_unused parameters, their grads should be kept as being
+    // undefined.
```

Comment:
This change is complicated enough that we probably want to add a NOTE in a single place which describes the overall logic in detail and then we can reference that note here.

Reply:
sounds good! will add notes